src/main/scala/chisel3/util/Counter.scala

```diff
@@ -59,4 +70,18 @@ object Counter
     when (cond) { wrap = c.inc() }
     (c.value, cond && wrap)
   }
+
+  /** Instantiate a [[Counter! counter]] with the specified number of counts, a gate, and a restart.
+   */
+  def apply(cond: Bool, n: Int, restart: Bool): (UInt, Bool) = apply(cond, n, restart, 0)
```

Comment:
Having 5 Counter constructors, each of which does quite different things, without a order of arguments that can be simply explained, probably isn't good for readability. 

If you look at code that uses this, can you immediately tell what each Counter invocation is doing and what the arguments mean? Especially if you were looking at code from months back?

Perhaps this would be an argument for Counter, CounterEnable, CounterReset, much like what we do with Regs?